Add first draft of how annotation types could look

cotramarko opened the pull request:
**Note: Don't merge this yet**

First draft of how annotation types could look and feel in the client. There are a couple of things I'm still a bit undecided on:

1. Right now I'm deserializing the annotation-types into a list of `AnnotationType(name:str)` objects. Is this the way to go, or just keep it as plain strings? 
2. I'm not a fan of `Optional[List[T]]`. But, with that said, I'm still not 100% into the current way I've replaced the `input_list_id=None` parameter with `annotation_types=[]` instead. Should this default to `None` instead of an empty list?

Ragge-dev commented on annotell-input-api/annotell/input_api/resources/abstract/creatable_input_api_resource.py:
```diff
@@ -17,26 +17,23 @@ def __init__(self, client: HttpClient, file_resource_client: FileResourceClient)
+                            input_request: dict,
+                            project: Optional[str],
+                            batch: Optional[str],
+                            annotation_types: List[str] = [],
```
Why `List[str]` though, if the API response is an `AnnotationType`?

Ragge-dev commented on annotell-input-api/annotell/input_api/resources/input/cameras.py:
```diff
@@ -14,27 +14,25 @@ def create(self,
                project: Optional[str] = None,
                batch: Optional[str] = None,
-               input_list_id: Optional[int] = None,
+               annotation_types: List[str] = [],
```
Aight, whatever the signature is here I think it should be the same as what we get from the API, otherwise it's confusing

fredrikjacobson commented:
I would argue for us to keep strings as the interface. We are not using tagged ids for `Project` or `Batch`. I get that we want to make it enumish, but there's no **real** type safety that will occur. Otherwise LGTM :)

cotramarko commented:
> 2. I have started to come around to your way of seeing this, in many cases it does make for cleaner code. However in python I would like to just ignore it and have it default to `None`, because having a parameter default to `[]` invites disaster and can lead to some unpredictable behavior. If **only** due to this I would like to have it as `Optional[List[AnnotationType]]`. Try this in the REPL and you'll see what I mean:
> 
> ```
> def func(a = []):
>     a.append(1)
>     return a
> ```
> 
> and call `func()` a couple of times

This will only be an issue if you mutate the argument in your function body, but in this case all we're ever doing is reading it.

Ragge-dev commented:
> > 1. I have started to come around to your way of seeing this, in many cases it does make for cleaner code. However in python I would like to just ignore it and have it default to `None`, because having a parameter default to `[]` invites disaster and can lead to some unpredictable behavior. If **only** due to this I would like to have it as `Optional[List[AnnotationType]]`. Try this in the REPL and you'll see what I mean:
> > 
> > ```
> > def func(a = []):
> >     a.append(1)
> >     return a
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and call `func()` a couple of times
> 
> This will only be an issue if you mutate the argument in your function body, but in this case all we're ever doing is reading it.

I would still argue it is inviting disaster even though it works today. We will have to change it anyway if we want to have a library without linting-errors since pylint complains about it as well.

cotramarko commented:
- Removed `AnnotationType` and instead opted for `str` everywhere.
- Added `Optional[List[T]]` as type-parameter for annotation_types with a default to `None` instead of empty list. 
- Added batch as optional parameter to get_annotation_types method

Ragge-dev reviewed (approved):
We just have to remember to also add the new batch route for annotation types before we push this to pypi.

Ragge-dev commented:
Also, don't forget to update the README.md

cotramarko commented:
Backend changes reflected in companion PR: https://github.com/annotell/annotell/pull/3099

cotramarko commented:
Scala deployed, merging this to master
